File: src/main/java/ru/octol1ttle/flightassistant/FlightAssistant.java
package ru.octol1ttle.flightassistant;

import net.fabricmc.api.ClientModInitializer;
import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.client.MinecraftClient;
import net.minecraft.util.Identifier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ru.octol1ttle.flightassistant.computers.ComputerHost;
import ru.octol1ttle.flightassistant.config.FAConfig;
import ru.octol1ttle.flightassistant.config.HUDConfig;
import ru.octol1ttle.flightassistant.hud.HudDisplayHost;

public class FlightAssistant implements ClientModInitializer {
    public static final String MODID = "flightassistant";
    public static final Logger LOGGER = LoggerFactory.getLogger("FlightAssistant");
    private static HudDisplayHost displayHost;
    private static ComputerHost computerHost;

    @Override
    public void onInitializeClient() {
        FAConfig.setup();
        FAKeyBindings bindings = new FAKeyBindings();
        bindings.registerAll();
        FACallbackListener.setup(new FACallbackListener(bindings));
    }

    public static void onClientStarted(MinecraftClient mc) {
        if (computerHost != null) {
            throw new IllegalStateException();
        }
        computerHost = new ComputerHost(mc);

        if (displayHost != null) {
            throw new IllegalStateException();
        }
        displayHost = new HudDisplayHost();
    }

    public static HudDisplayHost getDisplayHost() {
        if (displayHost == null) {
            throw new IllegalStateException();
        }
        return displayHost;
    }

    public static ComputerHost getComputerHost() {
        if (computerHost == null) {
            throw new IllegalStateException();
        }
        return computerHost;
    }

    public static Identifier id(String path) {
        return new Identifier(FlightAssistant.MODID, path);
    }

    public static boolean isHUDBatched() {
        return canUseBatching() && FAConfig.hud().batchedRendering != HUDConfig.BatchedRendering.NO_BATCHING;
    }

    public static boolean canUseBatching() {
        return FabricLoader.getInstance().isModLoaded("immediatelyfast");
    }
}

File: src/main/kotlin/ru/octol1ttle/flightassistant/FlightAssistant.kt
package ru.octol1ttle.flightassistant

import com.mojang.blaze3d.systems.RenderSystem
import dev.architectury.event.events.client.ClientLifecycleEvent
import net.minecraft.client.MinecraftClient
import net.minecraft.sound.SoundEvent
import net.minecraft.util.Identifier
import net.minecraft.util.math.RotationAxis
import org.joml.Matrix4f
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import ru.octol1ttle.flightassistant.api.event.FixedHudRenderCallback
import ru.octol1ttle.flightassistant.api.event.WorldRenderCallback
import ru.octol1ttle.flightassistant.api.util.RenderMatrices
import ru.octol1ttle.flightassistant.config.FAConfig
import ru.octol1ttle.flightassistant.impl.computer.ComputerHost
import ru.octol1ttle.flightassistant.impl.display.HudDisplayHost

object FlightAssistant  {
    const val MOD_ID: String = "flightassistant"
    internal val mc: MinecraftClient = MinecraftClient.getInstance()
    internal val logger: Logger = LoggerFactory.getLogger("FlightAssistant")

    internal fun init() {
        logger.info("Initializing (stage 1)")
        FAConfig.load()
        FAKeyBindings.setup()
        ClientLifecycleEvent.CLIENT_STARTED.register {
            logger.info("Initializing (stage 2)")
            HudDisplayHost.sendRegistrationEvent()
            ComputerHost.sendRegistrationEvent()
        }
        WorldRenderCallback.EVENT.register { tickDelta, camera, projectionMatrix, positionMatrix ->
            FAKeyBindings.checkPressed(ComputerHost)

            ComputerHost.tick(tickDelta)

            RenderMatrices.projectionMatrix.set(projectionMatrix)
            RenderMatrices.worldSpaceMatrix.set(positionMatrix)
            RenderMatrices.modelViewMatrix.set(RenderSystem.getModelViewMatrix())

            RenderMatrices.worldSpaceNoRollMatrix.set(Matrix4f().apply {
                rotate(RotationAxis.POSITIVE_X.rotationDegrees(camera.pitch))
                rotate(RotationAxis.POSITIVE_Y.rotationDegrees(camera.yaw + 180.0f))
            })

            RenderMatrices.ready = true
        }
        FixedHudRenderCallback.EVENT.register { context, _ ->
            HudDisplayHost.render(context)
        }
    }

    internal fun id(path: String): Identifier {
//? if >=1.21 {
        /*return Identifier.of(MOD_ID, path)
*///?} else
        return Identifier(MOD_ID, path)
    }

    internal fun soundEvent(name: String): SoundEvent {
        return SoundEvent.of(id(name))
    }
}
